File: Top_100_Liked_Questions/238._Product_of_Array_Except_Self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1] * len(nums)
        
        cumul = 1
        for i in range(len(nums)-1):
            cumul *= nums[i]
            ans[i+1] *= cumul
        
        cumul = 1
        for i in range(len(nums)-1, 0, -1):
            cumul *= nums[i]
            ans[i-1] *= cumul
        
        return ans

File: Top_100_Liked_Questions/238._Product_of_Array_Except_Self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];

        // [1, a, ab, abc, ...]
        ans[0] = 1;
        for (int i=0; i<nums.length-1; i++) {
            ans[i+1] = ans[i] * nums[i];
        }

        // [..., bcd, cd, d, 1]
        int mult_val = 1;
        for (int i=nums.length-1; i>0; i--) {
            mult_val *= nums[i];
            ans[i-1] *= mult_val;
        }

        return ans;
    }
}